app/origin_node.go

```diff
@@ -0,0 +1,65 @@
+package main
+
+import "github.com/weaveworks/scope/report"
+
+// OriginNode is a node in the originating report topology. It's a process ID
+// or network host. It's used by the /api/topology/{topology}/{nodeID} handler
+// to generate detailed information. One node from a rendered topology may
+// have multiple origin nodes.
+type OriginNode struct {
+	Table report.Table
+}
+
+func getOriginNode(t report.Topology, id string) (OriginNode, bool) {
+	node, ok := t.NodeMetadatas[id]
+	if !ok {
+		return OriginNode{}, false
+	}
+
+	// The node represents different actual things depending on the topology.
+	// So we deduce what it is, based on the metadata.
+	if _, ok := node["pid"]; ok {
+		return originNodeForProcess(node), true
+	}
+
+	// Assume network host. Could strengthen this guess by adding a
+	// special key in the probe spying procedure.
+	return originNodeForNetworkHost(node), true
+}
+
+func originNodeForProcess(node report.NodeMetadata) OriginNode {
+	rows := []report.Row{
+		{Key: "Host", ValueMajor: node["domain"], ValueMinor: ""},
+		{Key: "PID", ValueMajor: node["pid"], ValueMinor: ""},
+		{Key: "Process name", ValueMajor: node["name"], ValueMinor: ""},
+	}
+	if val, ok := node["docker_id"]; ok {
```

Comment:
Can do, but is it really better? I don't know. I kind of like it as-is, declaratively. But I'm a 2/10 on it, so if you feel more strongly, just say so and I change it.